src/Npgsql/NpgsqlParameterCollection.cs

```diff
@@ -748,10 +749,18 @@ internal void ProcessParameters(TypeMapper typeMapper, bool validateValues, Comm
     internal PlaceholderType PlaceholderType { get; set; }
 
     static NpgsqlParameter Cast(object? value)
-        => value is NpgsqlParameter p
-            ? p
-            : throw new InvalidCastException(
-                $"The value \"{value}\" is not of type \"{nameof(NpgsqlParameter)}\" and cannot be used in this parameter collection.");
+    {
+        var castedValue = value as NpgsqlParameter;
+        if (castedValue is null)
+            ThrowInvalidCastException(value);
+
+        return castedValue;
+    }
+
+    [DoesNotReturn]
+    static void ThrowInvalidCastException(object? value) =>
+        throw new InvalidCastException(
+            $"The value \"{value}\" is not of type \"{nameof(NpgsqlParameter)}\" and cannot be used in this parameter collection.");
```

Comment:
We could have a different exception message for nulls to be a bit clearer (but not super important).

Reply:
Just a different message like `NpgsqlParameterCollection doesn't accept null values` or also throw a different exception, as `ArgumentException`?

Reply:
Technically it's not exactly an argument (in the sense of a function argument)... But maybe that's still a good idea. Maybe check what SqlClient throws for that case?

(am ok with either...)